synapse/replication/tcp/client.py

```diff
@@ -131,8 +145,66 @@ async def on_rdata(
 
             await self.pusher_pool.on_new_notifications(token, token)
 
+        # Notify any waiting deferreds
+        waiting_list = self._streams_to_waiters.get(stream_name, [])
+
+        # Index of first item with a position after the current token, i.e we
+        # have called all deferreds before this index. If not overwritten by
+        # loop below means either a) no items in list so no-op or b) all items
+        # in list were called and so the list should be cleared. Setting it to
+        # `len(list)` works for both cases.
+        index_of_first_deferred_not_called = len(waiting_list)
+
+        for idx, (position, deferred) in enumerate(waiting_list):
+            if position <= token:
+                try:
+                    with PreserveLoggingContext():
+                        deferred.callback(None)
+                except Exception:
+                    # The deferred has been cancelled or timed out.
+                    pass
+            else:
+                index_of_first_deferred_not_called = idx
+                break
```

Comment:
We only want to call max one deferred per update?

Reply:
These are in order, so once we reach a deferred waiting for a position above the received row we know that *all* other entries have a position greater than the received row

Reply:
Unless the stream jumps ahead? Or is this presumed impossible?

Reply:
Turns out I thought we broke out of the loop upon running one of the deferreds, when that was not the case.